src/main/java/com/minecolonies/coremod/colony/colonyEvents/raidEvents/babarianEvent/BarbarianRaidEvent.java

```diff
@@ -0,0 +1,399 @@
+
+    /**
+     * The related colony
+     */
+    private IColony colony;
+
+    /**
+     * The events id
+     */
+    private int id;
+
+    /**
+     * The events starting spawnpoint
+     */
+    private BlockPos spawnPoint;
+
+    /**
+     * Status of the event
+     */
+    private EventStatus status = EventStatus.STARTING;
+
+    /**
+     * Days the event can last, to make sure it eventually despawns.
+     */
+    private int daysToGo = 3;
+
+    public BarbarianRaidEvent(IColony colony)
+    {
+        this.colony = colony;
+        id = colony.getEventManager().getAndTakeNextEventID();
+    }
+
+    @Override
+    public void setSpawnPoint(final BlockPos spawnPoint)
+    {
+        this.spawnPoint = spawnPoint;
+    }
+
+    @Override
+    public BlockPos getStartPos()
+    {
+        return spawnPoint;
+    }
+
+    @Override
+    public List<Entity> getEntities()
+    {
+        List<Entity> entities = new ArrayList<>();
+        entities.addAll(archers.keySet());
+        entities.addAll(chiefs.keySet());
+        entities.addAll(barbarians.keySet());
+        return entities;
+    }
+
+    @Override
+    public EventStatus getStatus()
+    {
+        return status;
+    }
+
+    @Override
+    public void setStatus(final EventStatus status)
+    {
+        this.status = status;
+    }
+
+    @Override
+    public int getID()
+    {
+        return id;
+    }
+
+    @Override
+    public ResourceLocation getEventTypeID()
+    {
+        return BABARIAN_RAID_EVENT_TYPE_ID;
+    }
+
+    @Override
+    public void setColony(@NotNull final IColony colony)
+    {
+        this.colony = colony;
+    }
+
+    @Override
+    public void registerEntity(final Entity entity)
+    {
+        if (!(entity instanceof AbstractEntityMinecoloniesMob))
+        {
+            entity.setDead();
+            return;
+        }
+
+        if (entity instanceof EntityChiefBarbarian && chiefs.keySet().size() < horde.numberOfBosses)
+        {
+            chiefs.put(entity, entity.getUniqueID());
+            return;
+        }
+
+        if (entity instanceof EntityArcherBarbarian && archers.keySet().size() < horde.numberOfArchers)
+        {
+            archers.put(entity, entity.getUniqueID());
+            return;
+        }
+
+        if (entity instanceof EntityBarbarian && barbarians.keySet().size() < horde.numberOfRaiders)
+        {
+            barbarians.put(entity, entity.getUniqueID());
+            return;
+        }
+
+        entity.setDead();
+    }
+
+    /**
+     * Called when an entity is removed
+     *
+     * @param entity
+     */
+    @Override
+    public void unregisterEntity(final Entity entity)
+    {
+        if (!(archers.containsKey(entity) || chiefs.containsKey(entity) || barbarians.containsKey(entity)) || status != EventStatus.PROGRESSING
+              || colony.getState() != ColonyState.ACTIVE)
+        {
+            return;
+        }
+
+        archers.remove(entity);
+        chiefs.remove(entity);
```

Comment:
why in this one here we just remove it from all and in the other ones we do it in the instanceof check?

Reply:
ye here its only needed to remove the entity from the list, while the other cases have special type-dependent logic, like reducing archer count, or when adding to only add to the right list